src/services/MetricsService/MetricsFilter.ts

```diff
@@ -0,0 +1,247 @@
+/* eslint-disable promise/no-nesting */
+import {getHoursBetween, getCurrentDate, daysBetweenUTC, getUTCMidnight} from 'shared/date-fns';
+import {ExposureStatus, ExposureStatusType} from 'services/ExposureNotificationService';
+
+import {DefaultMetricsFilterStateStorage, MetricsFilterStateStorage} from './MetricsFilterStateStorage';
+import {SecureKeyValueStore} from './SecureKeyValueStorage';
+
+export enum EventTypeMetric {
+  Installed = 'installed',
+  Onboarded = 'onboarded',
+  Exposed = 'exposed',
+  OtkNoDate = 'otk-no-date',
+  OtkWithDate = 'otk-with-date',
+  EnToggle = 'en-toggle',
+  ExposedClear = 'exposed-clear',
+  BackgroundCheck = 'background-check',
+  ActiveUser = 'active-user',
+  BackgroundProcess = 'background-process',
+}
+
+export type EventWithContext =
+  | {
+      type: EventTypeMetric.Installed;
+    }
+  | {
+      type: EventTypeMetric.Onboarded;
+    }
+  | {
+      type: EventTypeMetric.Exposed;
+    }
+  | {
+      type: EventTypeMetric.OtkNoDate;
+      exposureHistory: number[];
+    }
+  | {
+      type: EventTypeMetric.OtkWithDate;
+    }
+  | {
+      type: EventTypeMetric.EnToggle;
+      state: boolean;
+      onboardedDate: Date | undefined;
+    }
+  | {
+      type: EventTypeMetric.ExposedClear;
+      exposureStatus: ExposureStatus;
+    }
+  | {
+      type: EventTypeMetric.BackgroundCheck;
+    }
+  | {
+      type: EventTypeMetric.ActiveUser;
+    }
+  | {
+      type: EventTypeMetric.BackgroundProcess;
+      succeeded: boolean;
+      durationInSeconds: number;
+    };
+
+export interface FilteredEvent {
+  eventType: EventTypeMetric;
+  payload: [string, string][];
+  shouldBePushedToServerRightAway: boolean;
+}
+
+export interface MetricsFilter {
+  filterEvent(eventWithContext: EventWithContext): Promise<FilteredEvent | null>;
+  getDelayedOnboardedEventIfPublishable(
+    notificationStatus: string,
+    frameworkenabled: string,
+  ): Promise<FilteredEvent | null>;
+}
+
+export class DefaultMetricsFilter implements MetricsFilter {
+  private stateStorage: MetricsFilterStateStorage;
+
+  constructor(secureKeyValueStore: SecureKeyValueStore) {
+    this.stateStorage = new DefaultMetricsFilterStateStorage(secureKeyValueStore);
+  }
+
+  filterEvent(eventWithContext: EventWithContext): Promise<FilteredEvent | null> {
```

Comment:
Thinking about the naming here - addEvent seems like it might make more sense to someone calling this - yes it's filtered but does that matter - also in TypeScript we can see it returns a `FilteredEvent` which should help the dev know what's going on.

Reply:
`addEvent` is still the internal "public" API that will be used to publish events. `MetricsFilter` is a dependency of `FilteredMetricsService`